src/modules/commander/Commander.cpp

```diff
@@ -3888,203 +3829,183 @@ bool Commander::preflight_check(bool report)
 
-	for (int i = 0; i < ORB_MULTI_MAX_INSTANCES; i++) {
-		if (_telemetry[i].high_latency) {
-			high_latency_link_exists = true;
+				// handle different remote types
+				switch (telemetry.remote_type) {
+				case (telemetry_status_s::MAV_TYPE_GCS):
+					_datalink_last_heartbeat_gcs = telemetry.heartbeat_time;
 
-			if (status.high_latency_data_link_active) {
-				dl_loss_timeout_local = highlatencydatalink_loss_timeout;
-				dl_regain_timeout_local = highlatencydatalink_regain_timeout;
-
-			} else {
-				// if the high latency link is inactive we do not want to accidentally detect it as an active link
-				dl_loss_timeout_local = INT32_MIN;
-				dl_regain_timeout_local = INT32_MAX;
-			}
+					// Recover from data link lost
+					if (status.data_link_lost) {
+						if (hrt_elapsed_time(&_datalink_last_heartbeat_gcs) < (_datalink_regain_threshold.get() * 1_s)) {
+							status.data_link_lost = false;
+							status_changed = true;
+							mavlink_log_critical(&mavlink_log_pub, "DATA LINK REGAINED");
+						}
+					}
 
-		} else {
-			dl_loss_timeout_local = datalink_loss_timeout;
-			dl_regain_timeout_local = datalink_regain_timeout;
-		}
+					break;
 
-		if (_telemetry[i].last_heartbeat != 0 &&
-		    hrt_elapsed_time(&_telemetry[i].last_heartbeat) < dl_loss_timeout_local * 1e6) {
-			/* handle the case where data link was gained first time or regained,
-			 * accept datalink as healthy only after datalink_regain_timeout seconds
-			 * */
-			if (_telemetry[i].lost &&
-			    hrt_elapsed_time(&_telemetry[i].last_dl_loss) > dl_regain_timeout_local * 1e6) {
+				case (telemetry_status_s::MAV_TYPE_ONBOARD_CONTROLLER):
+					_datalink_last_heartbeat_onboard_controller = telemetry.heartbeat_time;
 
-				/* report a regain */
-				if (_telemetry[i].last_dl_loss > 0) {
-					mavlink_and_console_log_info(&mavlink_log_pub, "data link #%i regained", i);
+					if (_onboard_controller_lost) {
+						if (hrt_elapsed_time(&_datalink_last_heartbeat_onboard_controller) < _onboard_regain_threshold.get() * 1_s) {
+							mavlink_log_info(&mavlink_log_pub, "ONBOARD CONTROLLER REGAINED");
+							_onboard_controller_lost = false;
+						}
 
-				} else if (_telemetry[i].last_dl_loss == 0) {
-					/* new link */
-				}
+					}
 
-				/* got link again or new */
-				*status_changed = true;
+					if (telemetry.remote_component_id == telemetry_status_s::COMPONENT_ID_OBSTACLE_AVOIDANCE) {
+						if (telemetry.heartbeat_time != _datalink_last_heartbeat_avoidance_system) {
+							_avoidance_system_status_change = _datalink_last_status_avoidance_system != telemetry.remote_system_status;
+						}
 
-				_telemetry[i].lost = false;
-				have_link = true;
+						_datalink_last_heartbeat_avoidance_system = telemetry.heartbeat_time;
+						_datalink_last_status_avoidance_system = telemetry.remote_system_status;
 
-				if (!_telemetry[i].high_latency) {
-					have_low_latency_link = true;
-				}
+						if (_avoidance_system_lost != false) {
```

Comment:
```suggestion
						if (_avoidance_system_lost) {
```
Also please change the type of `_avoidance_system_lost` to `bool`.